### PR TITLE
Bump Eigen to 3.4.0

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Bumps Eigen to the 3.4 series. We could stick with 3.3.9 to fix
deprecated C++-17 warnings, but as there is only one breaking change
with 3.4.0 lets go all the way up in one single hit

**To test:**
- Check that all tests pass

There is no associated issue. - Fixes an issue seen on CI with newer MSVC compiler

*This does not require release notes* because it's not a user visible change

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by KedoKudo — DISMISSED** [review]
simple version bump, straight forward enough

**Comment by DavidFair** [comment]
@KedoKudo / @martyngigg Could you look at this again

The original commit wouldn't fix MSVC as it uses a pre-built. I've shifted that to now use the external project version rather than our third party deps repo

**Comment by DavidFair** [comment]
It looks like it's picking up the Qt4 backend on our OSX builders, any objections to me nuking the dependency from brew?

**Comment by martyngigg** [comment]
I think the Qt4 errors are warnings? I agree we don't need it but I'm a little loathed to fiddle with those machines too much.

The error seems to something about the root variable:

```
18:35:39 CMake Error at /usr/local/Cellar/cmake/3.17.3/share/cmake/Modules/CMakePackageConfigHelpers.cmake:265 (message):
18:35:39   Variable EIGEN_ROOT_DIR does not exist
```

Should we just upgrade the internal version inside the thirdparty instead for now?

**Comment by DavidFair** [comment]
Turns out it was a PEBKAC problem:
We were attempting to install Eigen to system directories on all platforms. OSX just does a better job protecting system dirs, hence the job failed

**Comment by DavidFair** [comment]
This should be ready to go now :crossed_fingers: 
My hope is by making it the same everywhere it makes it easier when we switch to Conda later on / upgrading it in the future too

File: buildconfig/CMake/Eigen.cmake [hunk 1]
```diff
@@ -1,30 +1,37 @@
...
+option(BUILD_TESTING OFF)
+option(EIGEN_LEAVE_TEST_IN_ALL_TARGET OFF)
+option(EIGEN_BUILD_PKGCONFIG OFF)
 
-if(WIN32)
-  # Installed by 3rd party dependencies bundle
-  set(USE_SYSTEM_EIGEN ON)
-endif()
+# Preserve shared attrs
+set(CMAKE_INSTALL_PREFIX_OLD ${CMAKE_INSTALL_PREFIX})
```
[the remark this answers is not in the document]
Comment (DavidFair) [reply buildconfig/CMake/Eigen.cmake:17]: Both libs install to `/usr/local` on Linux. We don't call the install target on any of these 3 deps, but Eigen uses the install_directory to configure some variables (hence `EIGEN_ROOT_DIR` was unset) just in-case we do run the install target

Agree 100%, I'd like to drop span (as I'm hoping the Conda compilers will have C++-20 support) and move Eigen and GTest over
My hope is by bringing all platforms over to 3.4.0 it makes it trivial for us to simply migrate this dep over as all platforms are already running the same version. If long-term we were going to stick with the third-party repo and *Nix system libs I'd have done this in the MSVC deps repo instead and dropped this PR

**Review by martyngigg — APPROVED** [review]
I'm happy with this now. It builds for me locally.

Just one question but not something to hold this up. Is it possible to tell Eigen we don't want to look for Qt?

**Comment by DavidFair** [comment]
So grepping their repo: `EIGEN_TEST_NOQT`, though it only affects a doc and test for a Mandelbrot set that requires Qt4
Want me to throw another commit on to pass this in too?

**Review by martyngigg — CHANGES_REQUESTED** [review]
I think that would be good.

Are the options actually taking effect? I notice that the testing option is off so I would have expected that `find_package(Qt4)` not even to run?

**Comment by DavidFair** [comment]
Done, and it still checks for configuring the `demo` and `doc` targets
It's worth double checking our OSX build if it tries to skip Qt4, as I don't have this installed locally

**Comment by martyngigg** [comment]
Yeah it's not worth holding these fixes up any further. I don't quite understand what's going on but using an available Eigen build is a better long term solution anyway.

**Review by martyngigg — APPROVED** [review]

**Review by Pasarus — CHANGES_REQUESTED** [review]
Could you bump this in the places for conda as well please? mantid-developer yml files would be first, and second would be the conda-recipes, in the conda-recipes repo.

File: buildconfig/CMake/Eigen.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (Pasarus) [reply buildconfig/CMake/Eigen.cmake:18]: The conda compilers are for only Linux, I see few issues with moving to GCC 11 for example on Linux, we will be dependent on MSVC and XCode versions moving forward with C++ standard support.

**Review by martyngigg — CHANGES_REQUESTED** [review]
@thomashampson  raised the issue of the cppcheck suppression removal and I think we do need to understand that or as ever it will just get left.

The [manual](https://cppcheck.sourceforge.io/manual.pdf) says that once macros `#if` are introduced then cppcheck will behave differently. Are we now seeing behaviour where cppcheck is just skipping run code because it has either maxed out its config depth or it's not running through a branch of code?

I wondered what might happen if we say forced cppcheck to look like `gcc` and using that config?

**Comment by martyngigg** [comment]
I've had a brief look/think about this and a couple of things.

Firstly I spotted that we should probably add `-i "${CMAKE_BINARY_DIR}"` to the `cppcheck` call as this will ignore those files and they won't even be checked. That way we don't need to suppressions and it speeds up the check as it skips that directory entirely.

One downside of switching to `FetchContent` is that `Eigen3` is not not treated as a system include so we have to manually suppress the warnings. I can't find a way to fix this as it suggests that `FetchContent` is really for things you want to consider part of your project. Is `ExternalProject` not a better model here as it is external and when we get an imported target it comes with `-isystem` so we don't need to manually suppress the warnings?

The only other thing would then be the `USE_SYSTEM_EIGEN` flag - do we not need this when we build with conda as we really want to use that anyway?

**Comment by DavidFair** [comment]
> Firstly I spotted that we should probably add `-i "${CMAKE_BINARY_DIR}"` to the `cppcheck` call as this will ignore those files and they won't even be checked. That way we don't need to suppressions and it speeds up the check as it skips that directory entirely.

Good point, I'll open a separate PR for this in case this one stalls out [Link TODO]

> Is `ExternalProject` not a better model here as it is external and when we get an imported target it comes with `-isystem` so we don't need to manually suppress the warnings?

I ended up switching these because `ExternalProject` obtains the source at build time (I'd normally link to the CMake docs paragraph, but there's no useful in-line link on that page), whereas we need them at configure time. The old mechanism hacked around this by simply invoking CMake to perform a "build" of Eigen during the configuration. 
This adds time onto Windows as we now have spin up CMake 3 times (Mantid, Eigen Configure and Eigen build/download). Whereas setting it to FetchContent makes it available at configure time in the top level CMake directly.

I wasn't aware of that, we can do this by [overriding the properties](https://stackoverflow.com/a/64139281) after calling FetchContent, I'll try that fix and see if we get `WConversion`....

> The only other thing would then be the `USE_SYSTEM_EIGEN` flag - do we not need this when we build with conda as we really want to use that anyway?
Woops I read `SYSTEM_EIGEN` as not one we control, forgetting that Conda appears like a system dependency

File: buildconfig/CMake/Eigen.cmake [hunk 2]
```diff
@@ -1,30 +1,44 @@
...
+  fetchcontent_makeavailable(Eigen)
 
-  execute_process(
-    COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" -DCMAKE_SYSTEM_VERSION=${CMAKE_SYSTEM_VERSION} .
-    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/extern-eigen
-  )
-  execute_process(COMMAND ${CMAKE_COMMAND} --build . WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/extern-eigen)
+  # Mark target as a system include
+  get_target_property(EIGEN_INC_DIRS eigen INTERFACE_INCLUDE_DIRECTORIES)
+  set_target_properties(eigen PROPERTIES INTERFACE_SYSTEM_INCLUDE_DIRECTORIES "${EIGEN_INC_DIRS}")
```
Comment (martyngigg) [inline buildconfig/CMake/Eigen.cmake:33]: Cool. Nice that we can keep the benefits of not running cmake again and again but keeping the system include 👍🏻

**Review by martyngigg — APPROVED** [review]
I'm happy with the changes here as they now only affect how we grab Eigen but have improved the performance of that mechanism on Windows.

**Comment by martyngigg** [comment]
@Pasarus Could you check the conda stuff here and then we could get this merged.

**Review by Pasarus — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Could I get this merged?

**Pull request merged**